src/algebra/ring_quot.lean

```diff
@@ -0,0 +1,272 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import ring_theory.ideal.basic
+import ring_theory.algebra
+
+/-!
+# Quotients of non-commutative rings
+
+Unfortunately, ideals have only been developed in the commutative case as `ideal`,
+and it's not immediately clear how one should formalise ideals in the non-commutative case.
+
+In this file, we directly define the quotient of a semiring by any relation,
+by building a bigger relation that represents the ideal generated by that relation.
+
+We prove the universal properties of the quotient,
+and recommend avoiding relying on the actual definition!
+
+Since everything runs in parallel for quotients of `R`-algebras, we do that case at the same time.
+-/
+
+universes u₁ u₂ u₃ u₄
+
+variables {R : Type u₁} [semiring R]
+variables (S : Type u₂) [comm_semiring S]
+variables {A : Type u₃} [semiring A] [algebra S A]
+
+namespace ring_quot
+
+inductive rel (r : R → R → Prop) : R → R → Prop
+| of ⦃x y : R⦄ (h : r x y) : rel x y
+| add_left ⦃a b c⦄ : rel a b → rel (a + c) (b + c)
+| mul_left ⦃a b c⦄ : rel a b → rel (a * c) (b * c)
+| mul_right ⦃a b c⦄ : rel b c → rel (a * b) (a * c)
+
+theorem rel.add_right {r : R → R → Prop} ⦃a b c : R⦄ (h : rel r b c) : rel r (a + b) (a + c) :=
+by { rw [add_comm a b, add_comm a c], exact rel.add_left h }
+
```

Comment:
Reiterating a previous comment as a suggested change
```suggestion

theorem rel.neg {R : Type u₁} [ring R] {r : R → R → Prop} ⦃a b : R⦄ (h : rel r a b) : rel r (-a) (-b) :=
by simp only [neg_eq_neg_one_mul a, neg_eq_neg_one_mul b, rel.mul_right h]

theorem rel.smul {r : A → A → Prop} (k : S) ⦃a b : A⦄ (h : rel r a b) : rel r (k • a) (k • b) :=
by simp only [algebra.smul_def, rel.mul_right h]
```

Reply:
Haven't tried, but for consistency `k` can probably be in ` ⦃⦄` too.

Reply:
Ah, sorry I neglected this on the previous pass, my mistake.